orangecontrib/spectroscopy/io/perkinelmer.py

```diff
@@ -0,0 +1,471 @@
+                reader.start = stops[-1]
+                stops = stops[:-1]
+
+                while reader.start >= stops[-1]:
+                    stops = stops[:-1]
+
+            else:
+                block_id, block_size = reader.read(6,
+                                                   format="<Hi",
+                                                   expect_tuple=True)
+
+                stops.append(reader.start + block_size)
+
+        info.update(PerkinElmer.decode5104(reader.read(block_size)))
+
+        reader.start = stops[1]
+
+        while not reader.atEnd():
+            block_id, block_size = reader.read(6,
+                                               format="<Hi",
+                                               expect_tuple=True)
+
+            if block_id in decoders.keys():
+                decoded = decoders[block_id](reader.peek(block_size))
+
+                if isinstance(decoded, dict):
+                    info.update(decoded)
+
+                else:
+                    spectrum = decoded
+
+            reader.start += block_size
+
+        wavenumbers = np.linspace(info['min_wavelength'],
+                                  info['max_wavelength'],
+                                  info['n_points'])
+
+        datavals = np.array(spectrum)[None, ...]
+
+        domain = Domain([], None)
+        meta_data = Table.from_numpy(domain,
+                                     X=np.zeros((len(datavals), 0)))
+
+        meta_data.attributes = info
+
+        return wavenumbers, datavals, meta_data
+
+    def read_fsm(self):
+        f = open(self.filename, "rb")
+        data = f.read()
+        f.close()
+
+        reader = BlockReader(data)
+
+        meta = {
+            "signature": reader.read(4, format="utf-8"),
+            "description": reader.read(40, format="utf-8"),
+        }
+
+        decoders = {
+            5100: PerkinElmer.decode5100,
+            5104: PerkinElmer.decode5104,
+            5105: PerkinElmer.decode5105,
+        }
+
+        spectrum = []
+
+        while not reader.atEnd(6):
+            block_id, block_size = reader.read(6,
+                                               format="<Hi",
+                                               expect_tuple=True)
+
+            decoded = decoders[block_id](reader.read(block_size))
+
+            if isinstance(decoded, dict):
+                meta.update(decoded)
+            else:
+                spectrum.append(decoded)
+
+        wavenumbers = np.arange(meta['z_start'],
+                                meta['z_end'] + meta['z_delta'],
+                                meta['z_delta'])
+
+        datavals = np.squeeze(spectrum)
+
+        domain = Domain([], None)
+        meta_data = Table.from_numpy(domain,
+                                     X=np.zeros((len(datavals), 0)))
+
+        meta_data.attributes = meta
+
+        return datavals, wavenumbers, meta_data
+
+    def read_spectra(self):
+        if self.filename[-2:] == "sp":

```

Comment:
We haven't done this before (2 readers in one) but I don't have a strong preference.

Better to use `os.path.splitext[1]` than string mangling I think.

Reply:
So, by joining two readers, you make a list of readers smaller, which does make sense. But what you lose is the ability to enforce a reader if a file perhaps has a wrong extension. If that is not a problem, go forward with the code, but even in this case I would only enforce one of the readers, so have `if ... else` instead of `if .. elif`.

One thing to fix. What i find problematic is that this string comparison respects letter case. You must do `.lower()` before comparing.it.

Reply:
Yeah, I think we need to unify readers as much as possible. See https://github.com/biolab/orange3/issues/6929 -  I don't know what is the best way for doing this but currently we are going towards producing lists that don't fit a screen anymore.

I did the changes for the logic but a bit differently... Let me know how you like it.

Reply:
I don't see how making the list shorter will help any use case (except the gui problem). Ideally the user never even looks at that list, auto-matching works. The only time they need the list is when they need to override the auto-matching and pick a specific reader, in which case they need the whole list!

Reply:
Yes, it is a GUI problem. Unfortunately the automatic recognition is not working all that well. Especially for general extensions (csv, txt, dat, etc. <-we should make this a filename too) and extensions where there can be multiple formats inside the file (NeaSPEC txt files). So I think unifying the readers by categories that make sense would be a good thing. Ideally I would like to see instrument or company names only once in the list and from there handle things well for each group. As we hopefully will support more and more file reading the list will become long in any case, but with this direction we can keep it more reasonable I think.

Reply:
Okay I think I understand. Some kind of middle ground, using more of the meta-reader infrastructure to shorten the list. And of course continuing to improve the automatic recognition!

If I was doing this one from scratch with your suggestion, I would do
PerkinElmerMeta -> PerkinElmerSP / PerkinElmerFSM but I think we've discussed the colour of this shed long enough :)

Reply:
Let's see what we can work out in the future. I'll open a new issue for the meta reader discussion.